### ISSUE TITLE
Improve ability to specify a TimeZone for formatting Dates/Times

### ISSUE THREAD

**Issue by twhitmorenz** [thymeleaf/thymeleaf#710]
Looking at the practicality of Date/Time formatting in ThymeLeaf. While some work has been done in https://github.com/thymeleaf/thymeleaf/issues/64 it seems more should be considered to align Date/Time formatting with common webapp requirements.

I'm using ThymeLeaf 2.1.6. But docs do not indicate anything changed in 3.0 in this regard. 

I'd like to propose some high-level requirements for a typical webapp:

- Appservers may frequently be located in a different location/timezone from the user. Because of this dates are stored without TZ, or in a standardized (not the user) TZ. Most simply the object model in memory will also hold them in this format (server TZ or without TZ).
- Server dates & times should be formatted appropriate to the user's timezone for display.
- Timezone for display should be provided by the webapp, potentially based on user- or browser-selected values.
- Most meaningfully, a single TZ can be used for rendering all dates/times on a page.

Based on some hours reading doc, trial and error, and Google searching my findings on how ThymeLeaf works seem to be:

- temporals.format() method cannot take a TimeZone parameter.
- There seems no method to provide a contextual TZ to be applied to rendering of all zoneless dates/times (eg. Instant) on the page.
- It looks possible to create a Calendar/DateTime incorporating a TZ by programmatically specifying every single component. This however looks extremely unwieldy as a means for formatting even a single date/time. Not sure if it's viable at all directly from an Instant. 
- Documentation seems limited in addressing the common requirements of implementing a webapp where the storage & model are in one TZ (or no TZ) and display should be rendered in user TZ.

Specific requests:
- Provide temporals.format(), dates.format() methods taking a ZoneId or TimeZone parameter.
- Look into providing a mechanism to contextualize the whole page as to TZ. Not sure what mechanisms are available, but this would be highly desirable; rather than having to specify a TZ or complex conversion on every single date output.
- Document how to localize dates in the view (as discussed here), rather than in the object model.

I've been working with ThymeLeaf for a short time only, using it with Spring Boot. Generally impressed and have been recommending it as a templating technology. Appreciate that I'm relatively new to ThymeLeaf and would welcome an easy way to do this; however have read the commit in #64 and can't yet see how it resolves the business requirements above. Would welcome any assistance!

**Comment by ultraq** [comment]
I agree with the first point of the feature request - being able to specify a time zone when working with temporals that don't have them can be useful.  Not sure how a "timezone for the current context" thing might work, but for the other Thymeleaf devs: if that were to be a thing, then is this discussion better suited over on the relevant module (https://github.com/thymeleaf/thymeleaf-extras-java8time)?

**Comment by jmiguelsamper** [comment]
I think the discussion belongs to both projects, #dates utility object belongs to Thymeleaf core and #temporals utility object to the Java8Time module. The idea is to keep both projects aligned, having the same methods and method signatures.

**Comment by danielfernandez** [comment]
As for the per-format-call timezone: that's already supported by `#temporals` in current versions.

As for the page-wide locale: that should be managed at the application level. In Spring, `LocaleContextHolder` could be used. Or even a `@ControllerAdvice` that injects into every view-layer execution the timeZone as a model attribute so that it is available to format expressions.

**Comment by OrangeDog** [comment]
I've found myself in an edge that the current formatting still doesn't support.
I have an `Instant`, which I want to format as a locale-aware date, but you have to convert it to a `LocalDate` first in order to use the locale-aware formats.

I've ended up having to define a model attribute for the timezone, as Thymeleaf doesn't support creating `ZoneId` instances, or alternatively a format of `MEDIUM_DATE` that can be applied to a full date/time.

```xml
<span th:text="${#temporals.format(time.atZone(utc).toLocalDate(), 'MEDIUM')}" />
```

**Comment by twhitmorenz** [comment]
Thanks @danielfernandez for the functionality, @OrangeDog for the follow-up. Timezone choice is inherently application-dependent (user preference, tenant config, request headers…), so I don't think Thymeleaf should resolve it itself — the cleanest approach is for the controller to supply the zone, or better a converted LocalDate, as a model attribute. Note you shouldn't need to create a ZoneId in the template: the new #temporals.format(temporal, pattern, zoneId) overload takes the zone as a plain string.

That said, you may have found a gap: the zoneId parameter appears to be ignored for locale-aware style names, and an Instant may always map to ofLocalizedDateTime. 

Is there a working way to format just date-only with a locale-aware style & a timezone? Something like ${#temporals.format(time, 'MEDIUM_DATE', zone)} seems like it could be a consistent extension — probably worth a separate follow-up issue since this one is closed.
